### PR TITLE
Use a random port for the test etcd server

### PR DESCRIPTION
**Pull request by ixdy**
This allows us to run multiple instances of the verification checks, the integration test, etc. simultaneously on a single host (e.g. the PR Jenkins).

ref #14781

@kubernetes/goog-testing

### CODE REVIEW

File: cmd/integration/integration.go [hunk 1]
```diff
@@ -955,6 +959,7 @@ type testFunc func(*client.Client)
 func addFlags(fs *pflag.FlagSet) {
 	fs.IntVar(
 		&maxConcurrency, "max-concurrency", -1, "Maximum number of tests to be run simultaneously. Unlimited if set to negative.")
+	fs.StringVar(&etcdServer, "etcd-server", "", "etcd server URL.")
```
Comment (wojtek-t) [inline cmd/integration/integration.go]: this also needs port, right? (maybe ... URL in http://host:port format)?

Also, how about being consistent with other places and use "etcd-servers" instead to handle multiple etcd servers too:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L214

**Comment by wojtek-t** [comment]
Just one minor comment - other than that LGTM.

File: hack/lib/etcd.sh [hunk 2]
```diff
@@ -39,22 +40,29 @@ kube::etcd::start() {
 
   # Start etcd
   ETCD_DIR=$(mktemp -d 2>/dev/null || mktemp -d -t test-etcd.XXXXXX)
-  kube::log::info "etcd -data-dir ${ETCD_DIR} --bind-addr ${host}:${port} >/dev/null 2>/dev/null"
-  etcd -data-dir ${ETCD_DIR} --bind-addr ${host}:${port} >/dev/null 2>/dev/null &
+  kube::log::info "Starting etcd with data-dir ${ETCD_DIR} and bind-addr ${ETCD_HOST}:${ETCD_PORT}"
+  etcd -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} >/dev/null 2>/dev/null &
```
Comment (lavalamp) [inline hack/lib/etcd.sh]: If it fails, retry with different port?

Comment (lavalamp) [inline hack/lib/etcd.sh]: I think you need to change the negotiation port, too? (the 7001 port)

File: cmd/integration/integration.go (hunk 1 above)
Comment (lavalamp) [inline cmd/integration/integration.go]: Or make it consistent with the other integration tests... weird to use a flag in one place and env vars in the other.

**Comment by lavalamp** [comment]
cc @timothysc

**Comment by timothysc** [comment]
@ixdy this affects the integration tests but do we run the UTs concurrently at all?

**Comment by lavalamp** [comment]
Yes, we run unit tests concurrently (but before the integration tests).

The tests in test/integration are also possibly run concurrently with each
other?

On Wed, Oct 14, 2015 at 9:53 AM, Timothy St. Clair <notifications@github.com

**Comment by timothysc** [comment]
@lavalamp, my mods are available and starts up an etcd on a random port.  imho we could probably leverage in the integration tests as well...

**Comment by ixdy** [comment]
The local etcd is started in a number of tests:
```
hack/local-up-cluster.sh:    kube::etcd::start
hack/test-update-storage-objects.sh:kube::etcd::start
hack/after-build/update-swagger-spec.sh:kube::etcd::start
hack/test-cmd.sh:kube::etcd::start
hack/test-integration.sh:  kube::etcd::start
hack/benchmark-integration.sh:  kube::etcd::start
```

My main focus here was on `update-swagger-spec.sh` (used in the verification scripts), `test-cmd.sh`, and `test-integration.sh`, since we'd like to start running all of those in the PR Jenkins instance, which runs several builds in parallel.

While the etcd stuff seems to work now in parallel, I've noticed that the integration test (in particular `cmd/integration`) still uses static ports, so the integration test is still not parallel-invocation-safe.

**Comment by timothysc** [comment]
imho #15392  should solve this issue, and randomizes the port properly to the clients to run in parallel.

The issue will be updating the integration tests to leverage.

**Comment by ixdy** [comment]
@timothysc will that also work for `hack/after-build/update-swagger-spec.sh`, which starts an apiserver and thus needs etcd?

**Comment by ixdy** [comment]
Closing since #16287 makes this unnecessary.

**Pull request closed without merging**